extensions/amp-vk/0.1/amp-vk.js

```diff
@@ -0,0 +1,249 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const embedTypes = {
+  POST: 'post',
+  POLL: 'poll',
+};
+
+
+import {Layout} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {removeElement} from '../../../src/dom';
+import {getData, listen} from '../../../src/event-helper';
+
+export class AmpVk extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?string} */
+    this.iframeUrl_ = null;
+
+    /** @private {?Object} */
+    this.iframeParams_ = null;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.iframePromise_ = null;
+
+    /** @private {?number} */
+    this.widgetHeight_ = 0;
+
+    /** @private {?Function} */
+    this.unlistenMessage_ = null;
+
+    /** @private {?string} */
+    this.embedType_ = null;
+
+    /** @private {?string} */
+    this.ownerId_ = null;
+
+    /** @private {?string} */
+    this.postId_ = null;
+
+    /** @private {?string} */
+    this.hash_ = null;
+
+    /** @private {?String} */
+    this.apiId_ = null;
+
+    /** @private {?String} */
+    this.pollId_ = null;
+  }
+
+  /**
+   * @param {boolean=} opt_onLayout
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    this.preconnect.url('https://vk.com', opt_onLayout);
+  }
+
+  /** @private */
+  getIFrameSrc_() {
+    const startWidth = this.element./*OK*/offsetWidth;
+    const pageUrl = this.element.ownerDocument
+        .location.href.replace(/#.*$/, '');
+    const pageReferrer = this.element.ownerDocument.referrer;
```

Comment:
@lannka just like to confirm that it is okay to sent referrer to a 3P component without page author specifying it (I assumed "yes" since other 3Ps have access to it through context)

Reply:
understood, switched to using getReferrerUrl() promise, also separated Poll and Post iframeSrc creation as we are planning to add more embed widgets in the future